cpp/src/snmg/utils.cuh

```diff
@@ -112,11 +110,81 @@ void allgather (SNMGinfo & env, size_t* offset, val_t* x_loc, val_t ** x_glob) {
 
 }
 
-void print_mem_usage()
-{
-  size_t free,total;
-  cudaMemGetInfo(&free, &total);  
-  std::cout<< std::endl<< "Mem used: "<<total-free<<std::endl;
+/**
+ * @tparam val_t The value type
+ * @tparam func_t The reduce functor type
+ * @param length The length of each array being combined
+ * @param x_loc Pointer to the local array
+ * @param x_glob Pointer to global array pointers
+ * @return Error code
+ */
+template <typename val_t, typename func_t>
+gdf_error treeReduce(SNMGinfo& env, size_t length, val_t* x_loc, val_t** x_glob){
+  auto i = env.get_thread_num();
+  auto p = env.get_num_threads();
+  env.setup_peer_access();
+  int rank = 1;
+  while(rank < p){
+    // Copy local data to the receiver's global buffer
+    if((i - rank) % (rank * 2) == 0){
+      int receiver = i - rank;
+      cudaMemcpyPeer(x_glob[receiver], receiver, x_loc, i, length*sizeof(val_t));

```

Comment:
Missing `cudaCheckError();` after cuda and thrust calls in this file

Reply:
I have added cudaCheckError() calls needed.